bsondump/bsondump_test.go

```diff
@@ -8,172 +8,284 @@ package bsondump
+	bufRef := new(bytes.Buffer)
+	bufRef.ReadFrom(outReference)
+	bufRefStr := bufRef.String()
+
+	// Get the output from a file.
+	outDump, err := os.Open(outFile)
+	require.NoError(err)
+	bufDump := new(bytes.Buffer)
+	bufDump.ReadFrom(outDump)
+	bufDumpStr := bufDump.String()
+
+	require.Equal(bufRefStr, bufDumpStr)
+}
+
+func testFromFileWithPositionalArgumentToFile(t *testing.T) {
+	require := require.New(t)
 
-	Convey("Test bsondump reading from a file with a positional arg and writing to a file", t, func() {
-		cmd := exec.Command(executable, "--outFile", "out.json", "testdata/sample.bson")
+	dir, cleanup := testutil.MakeTempDir(t)
+	defer cleanup()
+	outFile := filepath.Join(dir, "out.json")
 
-		err := cmd.Run()
-		So(err, ShouldBeNil)
+	cmd := bsondumpCommand("--outFile", outFile, "testdata/sample.bson")
 
-		// Get the correct bsondump result from a file to use as a reference.
-		outReference, err := os.Open("testdata/sample.json")
-		So(err, ShouldBeNil)
-		bufRef := new(bytes.Buffer)
-		bufRef.ReadFrom(outReference)
-		bufRefStr := bufRef.String()
+	err := cmd.Run()
+	require.NoError(err)
 
-		// Get the output from a file.
-		outDump, err := os.Open("out.json")
-		So(err, ShouldBeNil)
-		bufDump := new(bytes.Buffer)
-		bufDump.ReadFrom(outDump)
-		bufDumpStr := bufDump.String()
+	// Get the correct bsondump result from a file to use as a reference.
+	outReference, err := os.Open("testdata/sample.json")
+	require.NoError(err)
+	bufRef := new(bytes.Buffer)
+	bufRef.ReadFrom(outReference)
+	bufRefStr := bufRef.String()
 
-		So(bufDumpStr, ShouldEqual, bufRefStr)
+	// Get the output from a file.
+	outDump, err := os.Open(outFile)
+	require.NoError(err)
+	bufDump := new(bytes.Buffer)
+	bufDump.ReadFrom(outDump)
+	bufDumpStr := bufDump.String()
 
-		Reset(func() {
-			os.Remove("out.json")
-		})
+	require.Equal(bufRefStr, bufDumpStr)
+}
+
+func bsondumpCommand(args ...string) *exec.Cmd {
+	cmd := []string{"go", "run", filepath.Join("..", "bsondump", "main")}
+	cmd = append(cmd, args...)
+	return exec.Command(cmd[0], cmd[1:]...)
+}
+
+func TestBsondumpMaxBSONSize(t *testing.T) {
+	testtype.SkipUnlessTestType(t, testtype.UnitTestType)
+
+	// 16mb
+	size := int(16*math.Pow(1024, 2)) -
+		// subtract 5 for the 4 bytes for the int32 at the head of the document that
+		// specifies its size and 1 for its trailing NULL.
+		5 -
+		// subtract 1 byte for each byte that specifes the type of our two
+		// fields
+		2 -
+		// subtract 1 byte for the binary field subtype specifier
+		1 -
+		// subtract the length of each key name
+		len("name") -
+		len("content") -
+		// subtract 1 byte for the trailing NULL in each of our two keys
+		2 -
+		// subtract another 4 bytes for the int32 specifying the length of the
+		// each of each of our two fields
+		8 -
+		// subtract 1 for the string's trailing NULL
+		1
+
+	t.Run("bsondump with file at exactly max size of 16mb + 16kb", func(t *testing.T) {
+		_, err := runBsondumpWithLargeFile(t, size, true)
```

Comment:
The fact that `runBsondumpWithLargeFile(t, size, true)` tests a document of size `size` + 16kb seems a bit unintuitive and not to easy to follow. I think it would be better if it tests a doc of size `size` and we pass in (16 MB + 16 KB) or (16 MB + 16 KB + 1 B) instead.

Without looking at the implementation of `runBsondumpWithLargeFile()` it looked like the limit was still 16 MB.

Reply:
Yes, this is super confusing. I fixed it.